### ISSUE TITLE
Reporting of NaNs from getState changed in 7.5.1

### ISSUE THREAD

**Issue by dwhswenson** [openmm/openmm#3098]
In OpenPathSampling, we have a test that checks that we correctly handle failures when OpenMM generates NaNs, since [apparently NaNs are inevitable with single-precision forces and weak thermostats](https://github.com/openpathsampling/openpathsampling/pull/570#issuecomment-241212273). That test is failing as of OpenMM 7.5.1. It appears that the NaN coming from OpenMM, which we catch and handle internally, is now a `SystemError`. From previous behavior of OpenMM, we specifically check for NaNs by searching for `"nan"` and either `"particle"` or `"coordinates"` in the error message, and by explicitly checking the coordinates/velocities for NaNs (I think one of those behaviors dates back to circa OpenMM 6.1 and was changed in some later version).

Our test now raises the following error from `getState`:

```pytb
SystemError: <built-in function Context_getState> returned NULL without setting an error
```

You can see full details [here](https://github.com/openpathsampling/openpathsampling/runs/2480858567?check_suite_focus=true#step:9:1608) (until GitHub Actions deletes it). The stack trace is in OPS until we call `getState`, which raises the error when calling `_openmm.Context_getState`. Strangely, a related test passes on CI, but fails for me locally.

I'll put together a minimal example if you need it, but I thought I'd start by asking if this was intentional. My guess is that this is related to #3061, which looks like it changed exception reporting coming from `getState`.

**Comment by peastman** [comment]
Nothing related to this is supposed to have changed.  A minimal test case would be great!

**Comment by dwhswenson** [comment]
Here's a relatively minimal example. I've noticed that the following sometimes works in 7.5.1, too, but usually fails (which would be why I had 2 tests fail locally and only 1 fail on CI) -- so I'm even more convinced that this comes from #3061, since it seems like some kind of concurrency issue (which error I get is not deterministic).

Another thing I noticed: when 7.5.1 raises the NaN error, it comes from `LangevinIntegrator`, not from `getState` as with 7.5.0.

Input data here: [ala_small_traj.pdb.zip](https://github.com/openmm/openmm/files/6410059/ala_small_traj.pdb.zip)


```python
from simtk import openmm as mm
from simtk.openmm import app
from simtk import unit
import numpy as np

print(f"OpenMM: {mm.version.version}")

pdb = app.PDBFile("ala_small_traj.pdb")
forcefield = app.ForceField(
    'amber96.xml',  # solute FF
    'tip3p.xml'     # solvent FF
)

# OpenMM System
system = forcefield.createSystem(
    pdb.topology,
    nonbondedMethod=app.PME,
    nonbondedCutoff=1.0*unit.nanometers,
    constraints=app.HBonds,
    ewaldErrorTolerance=0.0005
)

integrator = mm.LangevinIntegrator(300*unit.kelvin,
                                   1.0/unit.picoseconds,
                                   2.0*unit.femtoseconds)
integrator.setConstraintTolerance(0.00001)

sim = mm.app.Simulation(topology=pdb.topology,
                        system=system,
                        integrator=integrator,
                        platform=mm.Platform.getPlatformByName('CPU'))
sim.context.setPositions(pdb.positions)

unit_vel = unit.nanometers / unit.picoseconds
velocities = np.zeros(shape=(system.getNumParticles(), 3))
velocities[0] = [1e6, 1e6, 1e6]
sim.context.setVelocities(velocities * unit_vel)

for _ in range(5):
    try:
        sim.step(2)
        state = sim.context.getState(getPositions=True,
                                     getVelocities=True,
                                     getEnergy=True)
    except Exception as e:
        if 'nan' in str(e):
            pass  # we can handle this
        else:
            raise  # we can't handle this

print("SUCCESS")
```

Outputs:

```
OpenMM: 7.5.0.dev-b49b82e
SUCCESS
```

```
OpenMM: 7.5.1.dev-a9cfd7f
Traceback (most recent call last):
  File "/Users/dwhs/tmp/openmm-7.5.1-nan/openmm-7.5.1-nan.py", line 42, in <module>
    state = sim.context.getState(getPositions=True,
  File "/Users/dwhs/miniconda3/envs/openmm-7.5.1/lib/python3.9/site-packages/simtk/openmm/openmm.py", line 2618, in getState
    state = _openmm.Context_getState(self, types, enforcePeriodicBox, groups_mask)
SystemError: <built-in function Context_getState> returned NULL without setting an error
```

**Comment by peastman** [comment]
I see the problem!  One of those stupid, silly mistakes.  In #3000 we renamed the top level Python module from `simtk.openmm` to just `openmm`.  As part of that, several lines in `exceptions.i` were modified.  For example

https://github.com/openmm/openmm/blob/a9cfd7fb9343e21c3dbb76e377c721328830a3ee/wrappers/python/src/swig_doxygen/swig_lib/python/exceptions.i#L33

changed to

https://github.com/openmm/openmm/blob/d8ef57fed6554ec95684e53768188e1f666405c9/wrappers/python/src/swig_doxygen/swig_lib/python/exceptions.i#L33

About six weeks later, #3061 added a similar block for `getState()`.  And of course it specified the module as `"openmm"`.  When we built 7.5.1, we included #3061 in it (since it's a bug fix) but not #3000.  And git happily merged the new block over to the old branch.

The result is that the block for `getState()` (but not the ones for `step()` or `minimize()`) has the wrong module name.  Most of the time you would never know.  The method still works perfectly.  But if it throws an exception, the handler tries to convert the C++ exception into a Python exception of class `openmm.OpenMMException`, which doesn't exist in that branch.  So the exception handler itself throws an exception.  Even that still basically works, but it means you now get an exception with the wrong message.

**Comment by dwhswenson** [comment]
Thanks for the explanation! And I hadn't been aware of dropping the old `simtk` -- I'm definitely in support of that.

**Comment by jchodera** [comment]
This would be a great exercise for making sure you're up for handling the whole conda-forge release cycle, @peastman!

**Comment by jchodera** [comment]
@dwhswenson : Thanks!
cc:  @mikemhenry who was looking into this! See https://github.com/openmm/openmm/issues/3098#issuecomment-838224141

**Comment by peastman** [comment]
I'm closing this now that 7.6 is released.  The problem never existed on the main branch, so there was nothing to fix.  It was created by a merge inconsistency on the 7.5 branch.

**Comment by dwhswenson** [comment]
Also, I can confirm that it was solved by the 7.6 RC (we had other issues to fix, but this one didn't resurface).
